pdr_backend/lake/table_bronze_pdr_slots.py

```diff
@@ -0,0 +1,152 @@
+from typing import Dict
+
+import polars as pl
+from enforce_typing import enforce_types
+from polars import Boolean, Float64, Int64, Utf8
+
+from pdr_backend.lake.plutil import (
+    pick_df_and_ids_on_period,
+)
+from pdr_backend.lake.table import Table
+from pdr_backend.ppss.ppss import PPSS
+
+
+bronze_pdr_slots_table_name = "bronze_pdr_slots"
+
+# CLEAN & ENRICHED PREDICTOOR SLOTS SCHEMA
+bronze_pdr_slots_schema = {
+    "ID": Utf8,  # f"{contract}-{slot}"
+    "contract": Utf8,  # f"{contract}"
+    "pair": Utf8,
+    "timeframe": Utf8,
+    "source": Utf8,
+    "slot": Int64,
+    "roundSumStakesUp": Float64,
+    "roundSumStakes": Float64,
+    "trueval": Boolean,
+    "timestamp": Int64,
+    "last_event_timestamp": Int64,
+}
+
+
+def _process_slots(
+    collision_ids: pl.Series, tables: Dict[str, Table], ppss: PPSS
+) -> Dict[str, Table]:
+    """
+    @description
+        Perform post-fetch processing on the data.
+        1. Find slots within the update
+        2. Transform slots to bronze
+        3. Concat to existing table
+    """
+
+    # only add new slots
+    slots_df = tables["pdr_slots"].df.filter(
+        (pl.col("timestamp") >= ppss.lake_ss.st_timestamp)
+        & (pl.col("timestamp") <= ppss.lake_ss.fin_timestamp)
+        & (pl.col("ID").is_in(collision_ids).not_())
+    )
+
+    # create contract column from the ID column
+    slots_df = slots_df.with_columns(
+        contract=pl.col("ID").map_elements(lambda s: s.split("-")[0])
+    )
+
+    if len(slots_df) == 0:
+        return tables
+
+    # Identify missing columns
+    missing_columns = [
+        col
+        for col in tables[bronze_pdr_slots_table_name].df.columns
+        if col not in slots_df.columns
+    ]
+
+    # Add missing columns to df2 using column-wise assignment
+    for col in missing_columns:
+        slots_df = slots_df.with_columns(
+            pl.Series(name=f"{col}", values=[None] * len(slots_df))
+        )
+
+    # Reorder the DataFrame columns based on the schema
+    slots_df = slots_df.select(bronze_pdr_slots_schema)
```

Comment:
I'm still confused by this code.
```
    # Identify missing columns
    missing_columns = [
        col
        for col in tables[bronze_pdr_slots_table_name].df.columns
        if col not in slots_df.columns
    ]

    # Add missing columns to df2 using column-wise assignment
    for col in missing_columns:
        slots_df = slots_df.with_columns(
            pl.Series(name=f"{col}", values=[None] * len(slots_df))
        )

    # Reorder the DataFrame columns based on the schema
    slots_df = slots_df.select(bronze_pdr_slots_schema)
```

What missing columns are you talking about?
It looks like you're joining a bunch of columns to a table and inserting "None" in a really inefficient way.

how to get an empty df with all required columns
```
slots_df = pl.DataFrame(schema=bronze_pdr_slots_schema)
```

Reply:
I don't see what is the issues here..
Why explicitly setting the columns helps? You have the bronze table schema defined above if you want to see the fields, this only adds them to slots table so that we can merge the slots values into our data frame.
We don't need to do any join in this step because the bronze_table is empty, what column should we join on?

Reply:
```Why explicitly setting the columns helps?```

Because whoever is reading the query/code can't just look at what you wrote and understand how it works.

In your original code, I can't tell what data is expected to be available when the slot is first created.. just that "there are a bunch of columns" that don't have the data.

With ETL/queries/dataframes, it's better to be explicit rather than implicit.

In your test, you're assuming that roundSumStakes and roundSumStakesUp are there when the slot is created... **but this is not how the data/events/system works.**

![image](https://github.com/oceanprotocol/pdr-backend/assets/69865342/608d4d1a-5864-4cbc-857d-536a0b34d3bc)